### PR TITLE
 Add File Downloading Capabilities

### PR DESCRIPTION
**Pull request by cryptidv**
### Background
This command will extend the power and functionality of Auto-GPT by allowing it to download files natively.

### Changes
* Added download_file command to allow Auto-GPT to download files natively, if needed.
* Added `--alow-downloads` flag

### Test Plan
Tested with GPT-3.5:
```
ai_goals:
- 'Download this file: "https://images.hdqwalls.com/wallpapers/cute-kitten-4k-im.jpg"
  and save it locally.'
ai_name: DownloaderGPT
ai_role: Downloads files. AND DOES NOTHING ELSE.
```
Downloaded the image successfully, I will further test with different file types, sizes etc.

Stress tested with:
```
ai_goals:
- 'Download this file: "https://download.manjaro.org/kde/22.0.5/manjaro-kde-22.0.5-230316-linux61.iso" and save it locally'
ai_name: DownloaderGPT
ai_role: Downloads files. AND DOES NOTHING ELSE.
```
### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

**Comment by cryptidv** [comment]
Why are so many imports being removed?

File: autogpt/file_operations.py [hunk 1]
```diff
@@ -139,3 +145,39 @@ def search_files(directory):
             found_files.append(relative_path)
 
     return found_files
+
+
+def download_file(url, filename):
+    """Downloads a file"""
```
Comment (harrymunro) [inline autogpt/file_operations.py]:
```suggestion
def download_file(url: str, filename: str):
    """Downloads a file

    Args:
        url (str): URL of the file to download
        filename (str): Filename to save the file as
    """
```

File: autogpt/spinner.py [hunk 2]
```diff
@@ -36,3 +37,10 @@ def __exit__(self, exc_type, exc_value, exc_traceback):
             self.spinner_thread.join()
         sys.stdout.write(f"\r{' ' * (len(self.message) + 2)}\r")
         sys.stdout.flush()
+
+    def update_message(self, new_message, delay=0.1):
+        """Update the spinner message"""
```
Comment (harrymunro) [inline autogpt/spinner.py]:
```suggestion
    def update_message(self, new_message: str, delay=0.1):
        """Update the spinner message

        Args:
            new_message (str): New message to display
            delay: Delay in seconds before updating the message
        """
```

File: autogpt/utils.py [hunk 3]
```diff
@@ -24,3 +24,12 @@ def validate_yaml_file(file: str):
         )
 
     return (True, f"Successfully validated {Fore.CYAN}`{file}`{Fore.RESET}!")
+
+
+def readable_file_size(size, decimal_places=2):
+    """Converts the given size in bytes to a readable format."""
```
Comment (harrymunro) [inline autogpt/utils.py]:
```suggestion
def readable_file_size(size, decimal_places: int=2):
    """Converts the given size in bytes to a readable format.
    
    Args:
        size: Size in bytes
        decimal_places (int): Number of decimal places to display
    """
```

**Review by harrymunro — COMMENTED** [review]
Suggested some docstring improvements and typing enforcement.

**Comment by andreas-it-dev-llc** [comment]
desperately waiting for this

**Review by nponeccop — CHANGES_REQUESTED** [review]
my bad

**Comment by nponeccop** [comment]
@cryptidv There are conflicts now

**Comment by cryptidv** [comment]
@nponeccop on it. Will create a new branch though, there are a crap-ton of conflicts.

**Comment by cryptidv** [comment]
@nponeccop @andreas-it-dev-llc closing this PR in favor of #1581

**Pull request closed without merging**